### ISSUE TITLE
[generic-worker] Ability to forward context from task definition to a config file

### ISSUE THREAD

**Issue by ahal** [taskcluster/taskcluster#6007]
*I'm filing this idea to get it down in writing, rather than for any particular business need.*

One issue that task authors face is how to pass context needed by a task down into the task itself. Currently there are two options:

1. Command line. Here's an example where we forward some of JSON-e context from a `.taskcluster.yml` into a Decision task:
https://github.com/taskcluster/taskgraph/blob/main/.taskcluster.yml#L248

2. Env. Here's an example where we JSONify a Python dict, stuff it into an env. The task then reads the env and converts it back to a dict:
https://searchfox.org/mozilla-central/source/taskcluster/gecko_taskgraph/transforms/job/mozharness_test.py#224

These options more or less work fine, though they can sometimes feel clunky. There can also be operating system limits to how large an environment variable can be, which can limit the amount of data that can be passed into the task.

A third option, might be for the workers to write a special part of the task definition to a config file in a known location. Then tasks can read the config file rather than read the environment. Scriptworkers actually do something similar, except they [write the entire task definition](https://github.com/mozilla-releng/scriptworker/blob/0a0e1e29bced3430752ef1994ff5c04225301101/src/scriptworker/context.py#L99) to a file. We could implement shared functions (perhaps in the tc clients?) for reading from this config file.

This feels like a less clunky way of passing down context into the tasks, and would avoid any limits to env length (though to my knowledge we haven't hit this yet).

**Comment by ahal** [comment]
Yeah that's a good idea actually, at least for the case of Taskgraph. We wouldn't be able to use this for the case of a `.tc.yml` spawning tasks though as there is no parent task to generate artifacts from.

**Comment by ahal** [comment]
Though another problem for that idea is that it would create a bottleneck. E.g, a Taskgraph Decision task can sometimes spawn thousands of tasks. It may be an issue if it needs to create thousands of artifacts, one for each task. I imagine it would slow down the decision task considerably. These artifacts would also be useless after the task finishes running, so it would likely add unnecessary storage costs as well.

**Comment by ahal** [comment]
Looks like Gitlab has a similar feature, except you can create `File` type variables which store the the value into a temporary file, and then the ENV points at that file:
https://docs.gitlab.com/ee/ci/variables/#use-file-type-cicd-variables

And just to re-iterate, this is not pressing :)

**Comment by petemoore** [comment]
Tasks can mount artifacts from previous tasks - is there a reason not to use the existing mounting mechanism? The decision task can write any artifacts it wants to, and then mount them in a generic-worker task.

**Comment by petemoore** [comment]
Note, they can also mount artifacts from an arbitrary url, with an optional sha256 hash for security.

**Comment by ahal** [comment]
Interesting, good to know! I think the fact most things are still `docker-worker` in Gecko is one reason I didn't know this..

I think the two drawbacks in https://github.com/taskcluster/taskcluster/issues/6007#issuecomment-1400533806 are still valid though. Namely, A) it'll slow down the decision task to write out 1000s of files and, B) it'll cost money to store the artifacts even though they'd be useless after the task runs.

**Comment by petemoore** [comment]
I see what you mean. I will give this some thought too. Off the bat, I don't feel like putting a huge amount of data in the task definition is great either, because task definitions aren't really designed to be very large. Ideally, if they require a large amount of data to operate, that is provided via some context link, such as a url, artifact, etc. When task definitions themselves become large unwieldy objects with a ton of data inside them, this can impact all sorts of things unexpectedly, like UI components that show task definitions, affect processing time of task definitions by the queue, perhaps database limits in storing the task definitions, etc.

However, I also see that you might be wary of generating thousands of artifacts, for the thousands of tasks that are generated. This concerns me less, because artifact expiration can be chosen appropriately, to stay around for as long as needed, but not longer. Would it be possible to organise such that the artifacts were spread over several tasks, rather than just the decision task? Or is it possible for some of the artifacts to be shared - e.g. in the most extreme case, a single artifact that contains the context for all the other tasks, and all tasks mount this artifact, but just use the config that applies to them, and ignore the rest? Perhaps the data is then not generated on a per-task dictionary fashion, but is normalised, leaving the responsibility with the task to extract the data it needs.

In general when a large set of data is needed in order to run a task, I feel like that should live outside the task definition and be fetched, to keep the task definition small and simple, where possible.
